debug_module.tex

```diff
@@ -227,13 +215,32 @@ \section{Abstract Commands} \label{abstractcommands}
     \end{tabulary}
 \end{table}
 
+\begin{commentary}
+    The Abstract Command interface is designed to allow a debugger to write
+    commands as fast as possible, and then later check whether they completed
+    without error.  In the common case the debugger will be much slower than
+    the target and commands succeed, which allows for maximum throughput. If
+    there is a failure, the interface ensures that no commands execute after
+    the failing one.  To discover which command failed, the debugger has to
+    look at the state of the DM (e.g.\ contents of \Rdatazero) or hart (e.g.\ 
+    contents of a register modified by a Program Buffer program) to determine
+    which one failed.
+\end{commentary}
+
 Before starting an abstract command, a debugger must ensure that \Fhaltreq,
 \Fresumereq, and \Fackhavereset are all 0.
 
 While an abstract command is executing (\Fbusy in \Rabstractcs is high), a
 debugger must not change \Fhartsel, and must not write 1 to \Fhaltreq,
 \Fresumereq, \Fackhavereset, \Fsetresethaltreq, or \Fclrresethaltreq.
 
+If a hart is reset while executing an abstract command then the abstract
+command is terminated, \Fbusy must go low, and \Fcmderr should be set to 6
+(interrupted).
+
+If an abstract command appears to be hung, a debugger can only regain control
+by resetting the hart.
```

Comment:
I assumed that the DM would notice that the hart reset (using havereset), and clean up whatever needs to be cleaned up on its side.

Reply:
No, I'm saying there may be other problems in the system (e.g. bus lock up) that means that even resetting the hart might not cause the debugger to regain control.

As a concrete example, in rocket-chip if the bus locks up (e.g. there's broken peripheral that just doesn't respond to bus transactions), potentially a hart can't execute any Program Buffer/Debug ROM instructions. Even if that hart was reset, it would still not be able to access those instructions.

Keep in mind that "hart reset" was an optional concept only added to the spec after a lot of debate, and was really only added there to save a bit for it. i don't think we should rely too heavily on the concept of "resetting a hart" for key debugger functionality.

Reply:
There are definitely problems that won't be solved by resetting a hart, but those kinds of problems shouldn't get fixed by resetting the DM. The DM should notice if one of its harts is reset, and if it's currently interacting with that hart, it needs to make sure to clean up state regarding to that interaction.